### ISSUE TITLE
XAPI's runtime dependency to gpumon, which requires a nVIDIA toolkit to build

### ISSUE THREAD

**Issue by stormi** [xapi-project/xen-api#4072]
Hi! I've reported it on IRC but I think it is good to have an open detailed issue about this here.

## Context
During the development phase of XCP-ng 8.1, I had (wrongly) decided to drop the `gpumon` package because I saw no way to build it using only open-source components. I also thought that it was only required for the nVIDIA VGPU features, who depend on the closed-source `vgpu` RPM, absent from XCP-ng. Although its source code is available and its license is FOSS-compatible (ISC License), building the binary requires a development toolkit from nVIDIA, whose license doesn't seem to be compatible with the legal requirements of the build system of an open project like XCP-ng.

## Issue
However, we found out the hard way (it went unnoticed during the beta and RC phases, unfortunately) that XAPI's startup phase would stall, without a visible error message, if an nVIDIA GPU is present and `gpumon` is missing.

In short, unless I'm mistaking about the licensing issues, XAPI currently has a runtime strong dependency to a non-free component.

There are other features from the XAPI that depend on closed-source components, but usually it will handle the absence of such a closed-source component gracefully. In the case of `gpumon`, though, the startup stalls.

## Expectation
I would expect that the Xen API be able to start and operate normally in the absence of `gpumon`, unless it turns out that my understanding of the build environment required to build it is wrong. It's not documented, so there may be too much guessing from my part.

Can we work together towards such a change? What would it involve?

**Comment by robhoes** [comment]
We should definitely make it so that `gpumon` is optional and its absence is gracefully handled.

**Comment by edwintorok** [comment]
gpumon has some code to handle the absence of the nvidia library at runtime, perhaps it needs to handle the absence of it at build time too.
Dune now has support for Configurator so it should be possible to check for the presence of the header and substitute either one or another .c/.ml file (where one .ml file would be a stub that always says unavailable)

**Comment by robhoes** [comment]
We could change the `gpumon_client.ml` in xcp-idl to check for gpumon, ideally by confirming that its message-switch queue exists. We'd need some need functionality in message-switch for that (we talked about this before, didn't we @edwintorok ?). Then we can return an error from the rpc function right away, rather than waiting forever, and xapi can handle this error.

**Comment by stormi** [comment]
For XCP-ng 8.2, we're trying a workaround that consists in building a "mock" gpumon, using `make mock`.

With that fake gpumon, XAPI is happy enough to start even if there's a nVIDIA GPU.

* Do you foresee possible issues with that setup, knowing that we don't provide vgpu support (`vgpu` RPM not included in the distribution)?
* How could we move forward so that this workaround is not needed anymore?

**Comment by stormi** [comment]
Today is the third anniversary for this issue. As we all agreed that gpumon should be made optional, do we have a chance of reaching this goal before the fourth one? :)

**Comment by psafont** [comment]
Christian did some work making gpumon load the library at runtime, I'm unsure whether this made it work better in this sense.
Tob had the intention to emrge gpumon into xen-api's repository. But hasn't had time top resolve this issue first
